src/common/debug.c

```diff
@@ -238,11 +237,16 @@ daos_key2str(daos_key_t *key)
 				break;
 			}
 		}
-		if (can_print) {
-			strncpy(buf, key->iov_buf, len);
-			buf[len] = 0;
-		} else {
-			strcpy(buf, "????");
+
+		if (can_print)
+			snprintf(buf, len, "%s", (char *)key->iov_buf);
```

Comment:
What will happen if:
key->iov_buf = "abcdefgh";
key->iov_buf_len = 8;
key->iov_len = 8;

The following "strlen(buf)" at line 248 may return invalid length that is larger than 32, then assert?

Reply:
So strnlen() may be safer?

Reply:
snprintf() always add \0 at the end. So, strlen() in 248 line should not return incorrect length.

Reply:
Probably we can remove line 227 and use here:
snprintf(buf, DF_KEY_STR_SIZE - 1, "%s", (char *)key->iov_buf);

Reply:
> 
@Nasf-Fan 
sorry I can't see the problem you described.

Reply:
Then the key string may be truncated under such case. In above example, the buf is "abcdefg", missed the last 'h'?

Reply:
Generally I'm for using strncpy over snprintf(..., "%s") where possible although they do have different semantics, is there a real issue here?

In the case above the string would not be NULL terminated, there are 8 characters in 8 bytes of space which is OK as long as we correctly check for it across the whole stack.

Reply:
>The following "strlen(buf)" at line 248 may return invalid length that is larger than 32, then assert?

The assertion checking and uint64 output will only be done if key->iov_len == 8, so strlen(buf) + 32 should always less than DF_KEY_STR_SIZE.

>Then the key string may be truncated under such case. In above example, the buf is "abcdefg", missed the last 'h'?

There are only 64 bytes buffer anyway, so it can only output 64 bytes at most. Sorry, I probably miss the question here, why last "h" is missed? the string is just 8 bytes?

Reply:
In my above example:

key->iov_buf = "abcdefgh";
key->iov_buf_len = 8;
key->iov_len = 8;

Then after snprintf(), it is expected the "buf" is the same as "key->iov_buf", but the real result is that the "buf" only contains the first 7 characters plus the terminated '\0'.

Reply:
So in this case we want non-terminated strings?

In the old code wasn't there a buffer-overflow in this case where buf[len] = 0; would write beyond the length of the buffer?

Reply:
If "buf" is not '\0' terminated, then the subsequent strlen() may get wrong result. In this patch, snprintf() will add a '\0' at the end of position (buf[len - 1]), so the original "abcdefgh" is truncated. I guess Di's idea is to copy whole "key->iov_buf" into the "buf" in spite of whether original "key->iov_buf" is '\0' terminated or not.

Reply:
The subsequent strlen() will, but there's also no reason to call strlen() at all, given we know what we've just written to the buffer, plus as per my other comment I'm not sure that writing two things to the buffer even makes sense.

Looking again at strncpy() vs snprintf() then only snprintf() returns the length written so that would be a case for using it over strncpy() in this case, if the real intent is to print both.

Reply:
In my example, snprintf() will return 8, although the 8th character is the '\0' instead of the expected 'h'.

Reply:
yes, you'd need to pass in "key->iov_len+1" as the length to strnprintf()

Reply:
Right, that is my concern.